File: app/src/main/java/com/mrboomdev/awery/extensions/data/CatalogMedia.java
package com.mrboomdev.awery.extensions.data;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.common.collect.Lists;
import com.mrboomdev.awery.util.Parser;

import java.io.Serial;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CatalogMedia implements Serializable {
	@Serial
	private static final long serialVersionUID = 1;
	@NonNull
	public String globalId;
	@Nullable
	public List<String> titles;
	@NonNull
	public Map<String, String> ids = new HashMap<>();
	@Nullable
	public Map<String, String> authors;
	@Nullable
	public String banner, description, country, ageRating, extra, url;
	@Nullable
	public MediaType type;
	@Nullable
	public ImageVersions poster = new ImageVersions();
	@Nullable
	public Calendar releaseDate;
	@Nullable
	public Integer duration, episodesCount, latestEpisode;
	@Nullable
	public Float averageScore;
	@Nullable
	public List<CatalogTag> tags;
	@Nullable
	public List<String> genres;
	@Nullable
	public MediaStatus status;

	/**
	 * @param globalId The unique id of the media in the following format:
	 * <p>{@code MANAGER_ID;;;PROVIDER_ID:EXTENSION_ID;;;ITEM_ID}</p>
	 */
	public CatalogMedia(@NonNull String globalId) {
		this.globalId = globalId;
	}

	public CatalogMedia(String managerId, String extensionId, String providerId, String mediaId) {
		this(managerId + ";;;" + providerId + ":" + extensionId + ";;;" + mediaId);
	}

	public CatalogMedia(@NonNull CatalogMedia original) {
		ageRating = original.ageRating;
		banner = original.banner;
		averageScore = original.averageScore;
		country = original.country;
		description = original.description;
		duration = original.duration;
		episodesCount = original.episodesCount;
		extra = original.extra;
		globalId = original.globalId;
		latestEpisode = original.latestEpisode;
		status = original.status;
		type = original.type;
		url = original.url;

		ids = new HashMap<>(original.ids);

		if(original.releaseDate != null) {
			releaseDate = Calendar.getInstance();
			releaseDate.setTimeInMillis(original.releaseDate.getTimeInMillis());
		}

		if(original.titles != null) {
			titles = List.copyOf(original.titles);
		}

		if(original.tags != null) {
			tags = List.copyOf(original.tags);
		}

		if(original.genres != null) {
			genres = List.copyOf(original.genres);
		}

		if(original.authors != null) {
			authors = Map.copyOf(original.authors);
		}

		if(original.poster != null) {
			poster = new ImageVersions(original.poster);
		}
	}

	public String getManagerId() {
		return globalId.split(";;;")[0];
	}

	public String getExtensionId() {
		return globalId.split(";;;")[1].split(":")[1];
	}

	public String getProviderId() {
		return globalId.split(";;;")[1].split(":")[0];
	}

	public String getId() {
		return globalId.split(";;;")[2];
	}

	@NonNull
	@Override
	public String toString() {
		return Parser.toString(CatalogMedia.class, this);
	}

	public void setTitle(@NonNull String... titles) {
		this.titles = Lists.newArrayList(titles);
	}

	@Nullable
	public String getTitle() {
		if(titles == null || titles.isEmpty()) return null;
		return titles.get(0);
	}

	public void setTitles(Collection<String> titles) {
		if(titles == null) {
			this.titles = new ArrayList<>();
			return;
		}

		this.titles = List.copyOf(titles);
	}

	@Nullable
	public String getBestBanner() {
		if(banner != null) return banner;
		return getBestPoster();
	}

	public void setId(String type, String id) {
		ids.put(type, id);
	}

	@Nullable
	public String getId(String type) {
		return ids.get(type);
	}

	public String getLargePoster() {
		if(poster != null && poster.large != null) {
			return poster.large;
		}

		return getBestPoster();
	}

	@Nullable
	public String getBestPoster() {
		if(poster == null) {
			return banner;
		}

		if(poster.extraLarge != null) return poster.extraLarge;
		if(poster.large != null) return poster.large;
		if(poster.medium != null) return poster.medium;
		return banner;
	}

	public void setPoster(String poster) {
		if(this.poster == null) {
			this.poster = new ImageVersions();
		}

		this.poster.extraLarge = poster;
		this.poster.large = poster;
		this.poster.medium = poster;
	}

	public enum MediaStatus {
		ONGOING, COMPLETED, COMING_SOON, PAUSED, CANCELLED, UNKNOWN
	}

	public enum MediaType {
		MOVIE, BOOK, TV, POST
	}

	public static class ImageVersions implements Serializable {
		@Serial
		private static final long serialVersionUID = 1;
		public String extraLarge, large, medium;

		public ImageVersions(@NonNull ImageVersions original) {
			extraLarge = original.extraLarge;
			large = original.large;
			medium = original.medium;
		}

		public ImageVersions() {}
	}
}

File: ext/src/main/kotlin/com/mrboomdev/awery/ext/data/CatalogMedia.kt
package com.mrboomdev.awery.ext.data
import com.mrboomdev.awery.ext.constants.AgeRating
import com.mrboomdev.awery.ext.util.GlobalId
import kotlinx.serialization.json.Json
import java.io.Serial
import java.io.Serializable

@kotlinx.serialization.Serializable
class CatalogMedia(
	/**
	 * A unique media id in the following format:
	 * MANAGER_ID;;;SOURCE_ID;;;MEDIA_ID
	 */
	val globalId: GlobalId,
	
	/**
	 * A title, which should be shown to the user.
	 */
	val title: String,
	
	/**
	 * An collection of other useful data such as trailer, screenshots, poster and banner not impactful on search
	 */
	val extras: Map<String, String> = emptyMap(),
	
	/**
	 * An set of ids for different services like for ex:
	 * ```json
	 * { "anilist": "57375", "mal": "473829" }
	 * ```
	 */
	val ids: Map<String, String> = emptyMap()
) : Serializable {
	
	constructor(original: CatalogMedia) : this(
		globalId = original.globalId,
		title = original.title,
		ids = original.ids.toMap(),
		extras = original.extras.toMap()
	)
	
	enum class Status {
		ONGOING, COMPLETED, COMING_SOON, PAUSED, CANCELLED
	}
	
	enum class Type {
		MUSIC, VIDEO, POST, MOVIE, BOOK
	}
	
	inline val status: Status?
		get() = extras[EXTRA_STATUS]?.uppercase()?.let {
			try { Status.valueOf(it) } catch(_: Throwable) { null } 
		}
	
	inline val type: Type?
		get() = when(extras[EXTRA_TYPE]?.lowercase()) {
			"movie", "tv" -> Type.MOVIE
			"book", "comic" -> Type.BOOK
			"post" -> Type.POST
			"video" -> Type.VIDEO
			"music" -> Type.MUSIC
			else -> null
		}
	
	inline val authors: Array<String>?
		get() = extras[EXTRA_AUTHORS]?.let {
			Json.decodeFromString(it)
		}
	
	inline val tags: Array<CatalogTag>?
		get() = extras[EXTRA_TAGS]?.let {
			Json.decodeFromString(it)
		}
	
	inline val genres: Array<String>?
		get() = extras[EXTRA_GENRES]?.let { 
			Json.decodeFromString(it)
		}
	
	inline val ageRating: AgeRating?
		get() = extras[EXTRA_AGE_RATING]?.let { AgeRating.match(it) }

	companion object {
		@Serial
		private val serialVersionUID = 2L
		
		/**
		 * @param poster [EXTRA_POSTER]
		 * @param banner [EXTRA_BANNER]
		 * @param episodes [EXTRA_EPISODES]
		 * @param description [EXTRA_DESCRIPTION]
		 * @param type [EXTRA_TYPE]
		 */
		@Suppress("NOTHING_TO_INLINE")
		inline fun create(
			globalId: GlobalId,
			title: String,
			ids: Map<String, String> = emptyMap(),
			description: String? = null,
			status: Status? = null,
			alternativeTitles: Array<String>? = null,
			type: String? = null,
			score: Int? = null,
			seasons: Int? = null,
			episodes: Int? = null,
			duration: Int? = null,
			latestEpisode: Int? = null,
			country: String? = null,
			trailer: String? = null,
			poster: String? = null,
			banner: String? = null,
			shareUrl: String? = null,
			reportUrl: String? = null,
			releaseDate: Long? = null,
			endDate: Long? = null,
			externalLinks: Array<ExternalLink>? = null,
			ageRating: AgeRating? = null,
			screenshots: Array<String>? = null,
			authors: Array<String>? = null,
			statusRatings: Array<Pair<Status, Long>>? = null,
			ratings: Array<Pair<Int, Long>>? = null,
			tags: Array<CatalogTag>? = null,
			genres: Array<String>? = null,
			extras: Map<String, String> = emptyMap()
		) = CatalogMedia(
			globalId = globalId,
			title = title,
			ids = ids,
			extras = buildMap {
				putSafe(EXTRA_TAGS, tags)
				putSafe(EXTRA_GENRES, genres)
				putSafe(EXTRA_AUTHORS, authors)
				putSafe(EXTRA_EXTERNAL_LINKS, externalLinks)
				putSafe(EXTRA_AGE_RATING, ageRating?.age)
				putSafe(EXTRA_END_DATE, endDate)
				putSafe(EXTRA_RELEASE_DATE, releaseDate)
				putSafe(EXTRA_REPORT, reportUrl)
				putSafe(EXTRA_SHARE, shareUrl)
				putSafe(EXTRA_BANNER, banner)
				putSafe(EXTRA_POSTER, poster)
				putSafe(EXTRA_COUNTRY, country)
				putSafe(EXTRA_TRAILER, trailer)
				putSafe(EXTRA_SCREENSHOTS, screenshots)
				putSafe(EXTRA_LATEST_EPISODE, latestEpisode)
				putSafe(EXTRA_DURATION, duration)
				putSafe(EXTRA_EPISODES, episodes)
				putSafe(EXTRA_SEASONS, seasons)
				putSafe(EXTRA_RATINGS, ratings)
				putSafe(EXTRA_STATUS_RATINGS, statusRatings)
				putSafe(EXTRA_SCORE, score)
				putSafe(EXTRA_TYPE, type)
				putSafe(EXTRA_ALT_TITLES, alternativeTitles)
				putSafe(EXTRA_STATUS, status?.name)
				putSafe(EXTRA_DESCRIPTION, description)
				putAll(extras)
			}
		)
		
		inline fun <reified T> MutableMap<String, String>.putSafe(key: String, value: T?) {
			value?.also { put(key, when(it) {
				is String -> it
				is Number -> it.toString()
				else -> Json.encodeToString(it)
			}) }
		}
		
		/**
		 * Current releasing status of the media. Allowed values:
		 * [Status.ONGOING], [Status.COMPLETED], [Status.COMING_SOON], [Status.PAUSED], [Status.CANCELLED]
		 */
		const val EXTRA_STATUS = "STATUS"
		
		/**
		 * An description of the media
		 */
		const val EXTRA_DESCRIPTION = "DESCRIPTION"
		
		/**
		 * Alternative names of the media in the json format:
		 * ["Name in German", "Name in Spanish"]
		 */
		const val EXTRA_ALT_TITLES = "ALT_TITLES"
		
		/**
		 * A type of the media. May be:
		 * ```
		 * "movie", "tv" -> Type.MOVIE
		 * "book", "comic" -> Type.BOOK
		 * "post" -> Type.POST
		 * "video" -> Type.VIDEO
		 * "music" -> Type.MUSIC
		 * ```
		 */
		const val EXTRA_TYPE = "TYPE"
		
		/**
		 * An average score from 0 to 100
		 * Displayed number may vary depending on the max value of [EXTRA_RATINGS]
		 */
		const val EXTRA_SCORE = "SCORE"
		
		/**
		 * Like [EXTRA_RATINGS], but for lists:
		 * ```
		 * [
		 *  [ "WATCHING", 393939934 ],
		 *  [ "COMPLETED", 3932 ],
		 *  [ "PLANNING, 932 ],
		 *  [ "DELAYED", 92999 ],
		 *  [ "DROPPED", 1000 ]
		 * ]
		 * ```
		 */
		const val EXTRA_STATUS_RATINGS = "STATUS_RATINGS"
		
		/**
		 * An json with statistics of each score:
		 * ```
		 * [
		 *  [ 5, 393939934 ],
		 *  [ 4, 3932 ],
		 *  [ 3, 932 ],
		 *  [ 2, 92999 ],
		 *  [ 1, 1000 ]
		 * ]
		 * ```
		 */
		const val EXTRA_RATINGS = "RATINGS"
		
		/**
		 * An origin country where this media been made
		 */
		const val EXTRA_COUNTRY = "COUNTRY"
		
		/**
		 * An average episode duration
		 */
		const val EXTRA_DURATION = "DURATION"
		
		/**
		 * Must be an url either on the video, either on the webpage with the video embed:
		 * https://example.com/trailer.mp4 or https://example.com/trailer_embed?video=483948&autoplay=true
		 */
		const val EXTRA_TRAILER = "TRAILER"
		
		/**
		 * An json array ["https://example.com/screenshot1.png", "https://example.com/screenshot2.png"]
		 */
		const val EXTRA_SCREENSHOTS = "SCREENSHOTS"
		
		/**
		 * An image url like:
		 * https://example.com/poster.png
		 */
		const val EXTRA_POSTER = "POSTER"
		
		/**
		 * An image url like:
		 * https://example.com/banner.png
		 */
		const val EXTRA_BANNER = "BANNER"
		
		/**
		 * A link, where user can go to report a bad content:
		 * https://example.com/report?media=388484
		 */
		const val EXTRA_REPORT = "REPORT"
		
		/**
		 * A link, which can be shared with friends to the content
		 * https://example.com/media/48493
		 */
		const val EXTRA_SHARE = "SHARE"
		
		/**
		 * A date when media was released (First episode) in time after epoch
		 * 393995858588494959
		 */
		const val EXTRA_RELEASE_DATE = "RELEASE_DATE"
		
		/**
		 * An json array of authors 
		 * ```json
		 * ["MrBoomDev", "MrBoomStudios"]
		 * ```
		 */
		const val EXTRA_AUTHORS = "AUTHORS"
		
		/**
		 * An json list of external services with the same media:
		 * ```json
		 * [
		 * 	{ "title": "Twitter", "link": "https://x.com/my_awesome_page", "icon": "https://example.com/twitter_icon.png" }
		 * ]
		 * ```
		 */
		const val EXTRA_EXTERNAL_LINKS = "EXTERNAL_LINKS"
		
		/**
		 * An amount of seasons
		 */
		const val EXTRA_SEASONS = "SEASONS"
		
		/**
		 * A total amount episodes
		 */
		const val EXTRA_EPISODES = "EPISODES"
		
		/**
		 * An amount of released episodes at the moment
		 */
		const val EXTRA_LATEST_EPISODE = "LATEST_EPISODE"
		
		/**
		 * An date when the latest episode was released
		 */
		const val EXTRA_END_DATE = "END_DATE"
		
		/**
		 * An age rating of the media. Users with lower age won't be able to access it
		 */
		const val EXTRA_AGE_RATING = "AGE_RATING"
		
		/**
		 * An json array of strings:
		 * ```
		 * ["action", "adventure", "horror"]
		 * ```
		 */
		const val EXTRA_GENRES = "GENRES"
		
		/**
		 * An json array of tags:
		 * ```
		 * [
		 * 	{ "name": "Ghosts", "isSpoiler": true }
		 * ]
		 * ```
		 */
		const val EXTRA_TAGS = "TAGS"
	}
}